src/control/lib/hardware/hwloc/provider.go

```diff
@@ -0,0 +1,278 @@
+//
+// (C) Copyright 2021 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+
+package hwloc
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/lib/hardware"
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+// NewProvider returns a new hwloc Provider.
+func NewProvider(log logging.Logger) *Provider {
+	return &Provider{
+		api: &api{},
+		log: log,
+	}
+}
+
+// Provider provides access to hwloc topology functionality.
+type Provider struct {
+	api *api
+	log logging.Logger
+}
+
+// GetTopology fetches a simplified hardware topology via hwloc.
+func (p *Provider) GetTopology(ctx context.Context) (*hardware.Topology, error) {
```

Comment:
I think it's fine for the moment, but I wonder if we should rework this a bit so that all providers follow the same pattern where they accept a `*hardware.Topology` and just decorate it as they know how. This would imply that the caller needs to instantiate the topology struct that gets passed around. Otherwise we have an ordering dependency and I don't think we want to lock ourselves into that.

Reply:
I was thinking about this as well. I agree.